plugins/embed-optimizer/hooks.php

```diff
@@ -23,90 +53,171 @@
+		add_action( 'wp_footer', 'embed_optimizer_lazy_load_scripts' );
+	}
+	return $html_processor->get_updated_html();
+}
 
-	/**
-	 * Determine how to lazy load the embed.
-	 *
-	 * - If there is only one iframe, set loading="lazy".
-	 * - Prevent making scripts lazy if there is an inline script.
-	 * - Only make script lazy if there is a single external script (since if there are
-	 *   multiple they may not get loaded in the right order).
-	 * - Ensure that both the iframe and the script are made lazy if both occur in the same embed.
-	 */
-	$iframe_count      = 0;
-	$script_count      = 0;
-	$has_inline_script = false;
-	// Locate the iframes and scripts.
-	while ( $html_processor->next_tag() ) {
-		if ( 'IFRAME' === $html_processor->get_tag() ) {
-			$loading_value = $html_processor->get_attribute( 'loading' );
-			if ( empty( $loading_value ) ) {
-				++$iframe_count;
-				if ( ! $html_processor->set_bookmark( 'iframe' ) ) {
-					wp_trigger_error( __FUNCTION__, esc_html__( 'Embed Optimizer unable to set iframe bookmark.', 'embed-optimizer' ) );
-					return $html;
+/**
+ * Applies changes to HTML in the supplied tag processor to lazy-load the embed.
+ *
+ * @since n.e.x.t
+ *
+ * @throws Exception But it won't because it is caught.
+ *
+ * @param WP_HTML_Tag_Processor|OD_HTML_Tag_Processor $html_processor HTML Processor.
+ * @return bool Whether the lazy-loading script is required.
+ */
+function embed_optimizer_update_markup( WP_HTML_Tag_Processor $html_processor ): bool {
+	$bookmark_names = array(
+		'script' => 'embed_optimizer_script',
+		'iframe' => 'embed_optimizer_iframe',
+	);
+	$trigger_error  = static function ( string $message ): void {
+		wp_trigger_error( __FUNCTION__, esc_html( $message ) );
+	};
+	try {
+		/*
+		 * Determine how to lazy load the embed.
+		 *
+		 * - If there is only one iframe, set loading="lazy".
+		 * - Prevent making scripts lazy if there is an inline script.
+		 * - Only make script lazy if there is a single external script (since if there are
+		 *   multiple they may not get loaded in the right order).
+		 * - Ensure that both the iframe and the script are made lazy if both occur in the same embed.
+		 */
+		$iframe_count      = 0;
+		$script_count      = 0;
+		$needs_lazy_script = false;
+		$has_inline_script = false;
+		$figure_depth      = 0;
+		// Locate the iframes and scripts.
+		do {
+			// This condition ensures that when iterating over an embed inside a larger document that we stop once we reach
+			// closing </figure> tag. The $processor is an OD_HTML_Tag_Processor when Optimization Detective is iterating
+			// over all tags in the document, and this embed_optimizer_update_markup() is usd as part of the tag visitor
+			// from Embed Optimizer. On the other hand, if $html_processor is not an OD_HTML_Tag_Processor then this is
+			// iterating over the tags of the embed markup alone as is passed into the embed_oembed_html filter.
+			if ( $html_processor instanceof OD_HTML_Tag_Processor ) {
```

Comment:
this feels a bit hacky to determine the context this way. could you add an additional parameter to `embed_optimizer_update_markup` instead (and then use a wrapper for the filter callback). not a blocker.

Reply:
OK, fair enough. Changed in 357364ed773140610c3d86babc4f7b93e551ed06.